File: components/gmaps-flow/version-1.0/google-map-flow/src/main/resources/META-INF/resources/frontend/google-map-polyline.js
/** 
@license
Copyright 2010-2019 Neotropic SAS <contact@neotropic.co>.

Licensed under the Apache License, Version 2.0 (the "License"); 
you may not use this file except in compliance with the License. 
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software 
distributed under the License is distributed on an "AS IS" BASIS, 
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
See the License for the specific language governing permissions and 
limitations under the License.
*/
import {html, PolymerElement} from '@polymer/polymer/polymer-element.js';
import * as Constants from './google-map-constants.js';
/**
 * @author Johny Andres Ortega Ruiz {@literal <johny.ortega@kuwaiba.org>}
 */
class GoogleMapPolyline extends PolymerElement {
	static get is() {
		return Constants.googleMapPolyline;
	}
	static get template() {
		return html`
			<style>
				:host {
					display: block;
				}
			</style>
		`;
	}
	static get properties() {
		return {
			draggable: {
				type: Boolean,
				value: false,
				observer: '_draggableChanged'
			},
			editable:{
				type: Boolean,
				value: false,
				notify: true,
				observer: '_editableChanged'
			},
			/**
			 * All CSS3 colors (except for extended named colors)
			 */
			strokeColor: {
				type: String,
				value: '#FF0000',
				observer: '_strokeColorChanged'
			},
			/**
			 * Between 0.0 and 1.0
			 */
			strokeOpacity: {
				type: Number,
				value: 1.0,
				observer: '_strokeOpacityChanged'
			},
			/**
			 * In pixels
			 */
			strokeWeight: {
				type: Number,
				value: 2,
				observer: '_strokeWeightChanged'
			},
			visible: {
				type: Boolean,
				value: true,
				observer: '_visibleChanged'
			},
			path: {
				type: Array,
				observer: '_pathChanged'
			},
			/**
			 * Indicates whether handles mouse events.
			 */
			clickable: {
				type: Boolean,
				value: true,
				observer: '_clickableChanged'
			},
			zIndex: {
				type: Number,
				observer: '_zIndexChanged'
			},
			label: {
				type: String,
				observer: '_labelChanged'
			},
			labelPosition: {
				type: Object, // {lat: 0.0, lng: 0.0}
				readOnly: true, // this._setLabelPosition({lat: 0.0, lng: 0.0})
				notify: true //label-position-changed
			},
			labelAnimation: {
				type: String,
				notify: true,
				observer: '_labelAnimationChanged'
			},
			labelIconUrl: {
				type: String
			},
      labelColor: {
        type: String,
				value: 'black',
				observer: '_labelColorChanged'
			},
			labelFontSize: {
				type: String,
				value: '14px',
				observer: '_labelFontSizeChanged'
			},
			labelClassName: {
				type: String,
				value: 'defaultLabel',
				observer: '_labelClassNameChanged'
			}
		};
	}
	/**
	 * @return {google.maps.MVCObject} The google.maps.Polyline
	 */
	getMVCObject() {
		return this.polyline;
	}		
	added(map) {
		var _this = this;
		this.polyline = new google.maps.Polyline({
			draggable: this.draggable,
			editable: this.editable,
			map: map,
			strokeColor: this.strokeColor,
			strokeOpacity: this.strokeOpacity,
			strokeWeight: this.strokeWeight,
			visible: this.visible,
			clickable: this.clickable
		});
		if (this.zIndex)
			this._zIndexChanged(this.zIndex);
		
		this.polyline.addListener('click', event => {
			if (event.vertex)
				this.dispatchEvent(new CustomEvent('vertex-click', {detail: {vertex: event.vertex}}));
			else
				this.dispatchEvent(new CustomEvent('polyline-click'));
		});
		this.polyline.addListener('dblclick', event => {
			if (event.vertex)
				this.dispatchEvent(new CustomEvent('vertex-dbl-click', {detail: {vertex: event.vertex}}));
			else
				this.dispatchEvent(new CustomEvent('polyline-dbl-click'));
		});
		this.polyline.addListener('mouseout', event => {
			this.dispatchEvent(new CustomEvent('polyline-mouse-out'));
		});
		this.polyline.addListener('mouseover', event => {
			this.dispatchEvent(new CustomEvent('polyline-mouse-over'));
		});
		this.polyline.addListener('rightclick', event => {
			if (event.vertex)
				this.dispatchEvent(new CustomEvent('vertex-right-click', {detail: {vertex: event.vertex}}));
			else
				this.dispatchEvent(new CustomEvent('polyline-right-click'));
		});
		this._setPolylinePath(this.path);

		if (this.label)
			this._labelChanged(this.label);
	}

	_setPolylinePath(path) {
		if (path.length >= 2) {
			var _this = this;
			this.polyline.setPath(path);
			
			google.maps.event.addListener(this.polyline.getPath(), 'insert_at', function(index) {
				_this._updatePath();
			});
			google.maps.event.addListener(this.polyline.getPath(), 'remove_at', function(index, removed) {
				_this._updatePath();
			});
			google.maps.event.addListener(this.polyline.getPath(), 'set_at', function(index, previous) {
				_this._updatePath();
			});
			this._setMarkerLabelPosition();
		}
	}

	_updatePath() {
		var path = [];
		this.polyline.getPath().forEach(function(coordinate, index) {
			path.push({lat: coordinate.lat(), lng: coordinate.lng()});
		});
		this.path = path;
		this._setMarkerLabelPosition();
		this.dispatchEvent(new CustomEvent('polyline-path-changed'));
	}

	removed() {
		if (this.polyline !== undefined)
			this.polyline.setMap(null);
		if (this.markerLabel)
			this.markerLabel.setMap(null);
		if (this.markerAnimation)
			this.markerAnimation.setMap(null);
	}

	_draggableChanged(newValue, oldValue) {
		if (this.polyline !== undefined && 
			this.polyline.getDraggable() !== newValue) {
			this.polyline.setDraggable(newValue);
		}
	}

	_editableChanged(newValue, oldValue) {
		if (this.polyline !== undefined && 
			this.polyline.getEditable() !== newValue) {
			this.polyline.setEditable(newValue);
		}
	}

	_strokeColorChanged(newValue, oldValue) {
		if (this.polyline !== undefined)
			this.polyline.setOptions({strokeColor: newValue});
	}

	_strokeOpacityChanged(newValue, oldValue) {
		if (this.polyline !== undefined)
			this.polyline.setOptions({strokeOpacity: newValue});
	}

	_strokeWeightChanged(newValue, oldValue) {
		if (this.polyline !== undefined)
			this.polyline.setOptions({strokeWeight: newValue});
	}

	_visibleChanged(newValue, oldValue) {
		if (this.polyline !== undefined && 
			this.polyline.getVisible() != newValue) {
			this.polyline.setVisible(newValue);
			if (this.markerLabel) {
				this.markerLabel.setVisible(newValue);
			}
		}
	}

	_pathChanged(newValue, oldValue) {		
		if (this.polyline !== undefined && 
			this.polyline.getPath() !== undefined) { 
			if (this.polyline.getPath().getLength() === this.path.length) {
				var flag = false;
				for (var i = 0; i < this.path.length; i++) {
					if (this.polyline.getPath().getAt(i).lat() != this.path[i].lat ||  
						this.polyline.getPath().getAt(i).lng() != this.path[i].lng) {
						flag = true;
						break;
					}
				}
				if (flag) {
					this._setPolylinePath(newValue);
				}
			} else {
				this._setPolylinePath(newValue);
			}
		}
	}
	/**
	 * Indicates whether handles mouse events.
	 * @param {boolean} newValue 
	 */
	_clickableChanged(newValue) {
		if (this.polyline)
			this.polyline.setOptions({clickable: newValue});
	}
	/**
	 * @param {number} newValue
	 */
	_zIndexChanged(newValue) {
		if (this.polyline)
			this.polyline.setOptions({zIndex: newValue});
	}
	_labelChanged(newValue) {
		if (this.markerLabel) {
			this.markerLabel.setLabel(newValue ? 
				{
					text: newValue,
					className: this.labelClassName ? this.labelClassName : 'defaultLabel',
					color: this.labelColor ? this.labelColor : 'black',
					fontSize: this.labelFontSize ? this.labelFontSize : '14px'
				} 
				: newValue
			);
		}
	}
	_setMarkerLabelPosition() {
		if (!this.markerLabel) {
			this.markerLabel = new google.maps.Marker({
				map: this.polyline.getMap(),
				clickable: false,
				icon: {
					path: google.maps.SymbolPath.CIRCLE,
					labelOrigin: new google.maps.Point(0, 20),
					strokeOpacity: 0
				}
			});

			this.markerAnimation = new google.maps.Marker({
				icon: this.labelIconUrl,
			});
			this.markerAnimation.addListener('click', event => this.labelAnimation = null);
		}
		var _path = this.polyline.getPath();
		var i = Math.floor(_path.getLength() / 2);

		var markerLabelPosition = null;
		if (_path.getLength() % 2 == 0) {
			markerLabelPosition = google.maps.geometry.spherical.interpolate(_path.getAt(i - 1), _path.getAt(i), 0.5);
		} else {
			markerLabelPosition = _path.getAt(i);
		}
		this._setLabelPosition({lat: markerLabelPosition.lat(), lng: markerLabelPosition.lng()});
		this.markerLabel.setPosition(markerLabelPosition);
		this.markerAnimation.setPosition(markerLabelPosition);
	}
	_labelAnimationChanged(newValue) {
		if (this.markerLabel) {
			if ('bounce' === newValue) {
				this.markerAnimation.setMap(this.polyline.getMap());
				this.markerAnimation.setAnimation(google.maps.Animation.BOUNCE);
			} else if ('drop' === newValue) {
				this.markerAnimation.setMap(this.polyline.getMap());
				this.markerAnimation.setAnimation(google.maps.Animation.DROP);
			} else {
				this.markerAnimation.setMap(null);
				this.markerAnimation.setAnimation(null);
			}
		}
	}
	_labelColorChanged(newValue) {
		if (this.markerLabel && this.markerLabel.getLabel() && this.markerLabel.getLabel().color !== newValue) {
			this.markerLabel.setLabel({
					text: this.label,
					className: this.labelClassName ? this.labelClassName : 'defaultLabel',
					color: newValue ? newValue : 'black',
					fontSize: this.labelFontSize ? this.labelFontSize : '14px'
			});
		}
	}
  _labelFontSizeChanged(newValue) {
    if (this.markerLabel && this.markerLabel.getLabel() && this.markerLabel.getLabel().fontSize !== newValue) {
			this.markerLabel.setLabel({
					text: this.label,
					className: this.labelClassName ? this.labelClassName : 'defaultLabel',
					color: this.labelColor ? this.labelColor : 'black',
					fontSize: newValue ? newValue : '14px'
			});
		}
  }
  _labelClassNameChanged(newValue) {
    if (this.markerLabel && this.markerLabel.getLabel() && this.markerLabel.getLabel().className !== newValue) {
			this.markerLabel.setLabel({
					text: this.label,
					className: newValue ? newValue : 'defaultLabel',
					color: this.labelColor ? this.labelColor : 'black',
					fontSize: this.labelFontSize ? this.labelFontSize : '14px'
			});
		}
  }
}

window.customElements.define(GoogleMapPolyline.is, GoogleMapPolyline);

File: components/gmaps-flow/version-1.0/google-map-flow/src/main/java/com/neotropic/flow/component/googlemap/GoogleMapPolyline.java
/*
 *  Copyright 2010-2019 Neotropic SAS <contact@neotropic.co>.
 *  
 *  Licensed under the Apache License, Version 2.0 (the "License"); 
 *  you may not use this file except in compliance with the License. 
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software 
 *  distributed under the License is distributed on an "AS IS" BASIS, 
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
 *  See the License for the specific language governing permissions and 
 *  limitations under the License.
 */
package com.neotropic.flow.component.googlemap;

import com.vaadin.flow.component.Component;
import com.vaadin.flow.component.ComponentEventListener;
import com.vaadin.flow.component.Synchronize;
import com.vaadin.flow.component.Tag;
import com.vaadin.flow.component.dependency.JsModule;
import com.vaadin.flow.shared.Registration;
import elemental.json.Json;
import elemental.json.JsonArray;
import elemental.json.JsonObject;
import java.util.List;

/**
 *
 * @author Johny Andres Ortega Ruiz {@literal <johny.ortega@kuwaiba.org>}
 */
@Tag("google-map-polyline")
@JsModule("./google-map-polyline.js")
public class GoogleMapPolyline extends Component {
    private List<LatLng> path;
    
    public GoogleMapPolyline() {
    }
    
    public boolean getDraggable() {
        return getElement().getProperty(Constants.Property.DRAGGABLE, Constants.Default.DRAGGABLE);
    }
    
    public void setDraggable(boolean draggable) {
        getElement().setProperty(Constants.Property.DRAGGABLE, draggable);
    }
    @Synchronize(property="editable", value="editable-changed")
    public boolean getEditable() {
        return getElement().getProperty(Constants.Property.EDITABLE, Constants.Default.EDITABLE);
    }
    
    public void setEditable(boolean editable) {
        getElement().setProperty(Constants.Property.EDITABLE, editable);
    }
    
    public String getStrokeColor() {
        return getElement().getProperty(Constants.Property.STROKE_COLOR, Constants.Default.STROKE_COLOR);
    }   
    
    public void setStrokeColor(String strokeColor) {
        getElement().setProperty(Constants.Property.STROKE_COLOR, strokeColor);
    }
    
    public double getStrokeOpacity() {
        return getElement().getProperty(Constants.Property.STROKE_OPACITY, Constants.Default.STROKE_OPACITY);
    }
    
    public void setStrokeOpacity(double strokeOpacity) {
        getElement().setProperty(Constants.Property.STROKE_OPACITY, strokeOpacity);
    }
    
    public double getStrokeWeight() {
        return getElement().getProperty(Constants.Property.STROKE_WEIGHT, Constants.Default.STROKE_WEIGHT);
    }
    
    public void setStrokeWeight(double strokeWeight) {
        getElement().setProperty(Constants.Property.STROKE_WEIGHT, strokeWeight);
    }
    
    public boolean getPolylineVisible() {
        return getElement().getProperty(Constants.Property.VISIBLE, Constants.Default.VISIBLE);
    }
    
    public void setPolylineVisible(boolean visible) {
        getElement().setProperty(Constants.Property.VISIBLE, visible);
    }
    /**
     * Indicates whether handles mouse events. Default value is true.
     * @return Indicates whether handles mouse events.
     */
    public boolean getClickable() {
        return getElement().getProperty(Constants.Property.CLICKABLE, true);
    }
    /**
     * Indicates whether handles mouse events.
     * @param clickable Indicates whether handles mouse events.
     */
    public void setClickable(boolean clickable) {
        getElement().setProperty(Constants.Property.CLICKABLE, clickable);
    }
    
    public double getZIndex() {
        return getElement().getProperty(Constants.Property.Z_INDEX, 0);
    }
    
    public void setZIndex(double zIndex) {
        getElement().setProperty(Constants.Property.Z_INDEX, zIndex);
    }
    
    @Synchronize(property="path", value="polyline-path-changed")
    public List<LatLng> getPath() {
        if (getElement().getPropertyRaw("path") instanceof JsonArray) {
            path.clear();
            JsonArray jsonCoordinates = (JsonArray) getElement().getPropertyRaw("path");
            for (int i = 0; i < jsonCoordinates.length(); i++) {
                JsonObject jsonCoordinate = jsonCoordinates.getObject(i);
                double lat = jsonCoordinate.getNumber("lat");
                double lng = jsonCoordinate.getNumber("lng");
                path.add(new LatLng(lat, lng));
            }
        }
        return path;
    }
    
    public void setPath(List<LatLng> path) {
        this.path = path;
        
        JsonArray jsonPath = Json.createArray();
        for (int i = 0; i < path.size(); i++) {
            JsonObject jsonCoordinate = Json.createObject();
            jsonCoordinate.put("lat", path.get(i).getLat());
            jsonCoordinate.put("lng", path.get(i).getLng());
            jsonPath.set(i, jsonCoordinate);
        }
        getElement().setPropertyJson(Constants.Property.PATH, jsonPath);
    }
    
    public String getLabel() {
        return getElement().getProperty(Constants.Property.LABEL, null);
    }
    
    public void setLabel(String label) {
        getElement().setProperty(Constants.Property.LABEL, label);
    }
    @Synchronize(property = "labelPosition", value = "label-position-changed")
    public LatLng getLabelPosition() {
        JsonObject labelPosition = (JsonObject) getElement().getPropertyRaw("labelPosition");
        return labelPosition != null ? new LatLng(labelPosition.getNumber("lat"), labelPosition.getNumber("lng")) : null;
    }
    @Synchronize(property = "labelAnimation", value = "label-animation-changed")
    public Animation getLabelAnimation() {
        return Animation.getAnimation(
            getElement().getProperty(Constants.Property.LABEL_ANIMATION, null)
        );
    }
    
    public void setLabelAnimation(Animation animation) {
        if (animation != null)
            getElement().setProperty(Constants.Property.LABEL_ANIMATION, animation.animation());
        else
            getElement().setPropertyJson(Constants.Property.LABEL_ANIMATION, Json.createNull());
    }
    
    public String getLabelIconUrl() {
        return getElement().getProperty(Constants.Property.LABEL_ICON_URL, null);
    }
    
    public void setLabelIconUrl(String labelIconUrl) {
        getElement().setProperty(Constants.Property.LABEL_ICON_URL, labelIconUrl);
    }
    
    public String getLabelColor() {
        return getElement().getProperty(Constants.Property.LABEL_COLOR, null);
    }
    
    public void setLabelColor(String labelColor) {
        getElement().setProperty(Constants.Property.LABEL_COLOR, labelColor);
    }
    
    public String getLabelFontSize() {
        return getElement().getProperty(Constants.Property.LABEL_FONT_SIZE, null);
    }
    
    public void setLabelFontSize(String labelFontSize) {
        getElement().setProperty(Constants.Property.LABEL_FONT_SIZE, labelFontSize);
    }
    
    public String getLabelClassName() {
        return getElement().getProperty(Constants.Property.LABEL_CLASS_NAME, null);
    }
    
    public void setLabelClassName(String labelClassName) {
        getElement().setProperty(Constants.Property.LABEL_CLASS_NAME, labelClassName);
    }
    
    public Registration addPolylineClickListener(ComponentEventListener<GoogleMapEvent.PolylineClickEvent> listener) {
        return addListener(GoogleMapEvent.PolylineClickEvent.class, listener);
    }
        
    public Registration addPolylineDblClickListener(ComponentEventListener<GoogleMapEvent.PolylineDblClickEvent> listener) {
        return addListener(GoogleMapEvent.PolylineDblClickEvent.class, listener);
    }
    
    public Registration addPolylineMouseOutListener(ComponentEventListener<GoogleMapEvent.PolylineMouseOutEvent> listener) {
        return addListener(GoogleMapEvent.PolylineMouseOutEvent.class, listener);
    }
    
    public Registration addPolylineMouseOverListener(ComponentEventListener<GoogleMapEvent.PolylineMouseOverEvent> listener) {
        return addListener(GoogleMapEvent.PolylineMouseOverEvent.class, listener);
    }
    
    public Registration addPolylineRightClickListener(ComponentEventListener<GoogleMapEvent.PolylineRightClickEvent> listener) {
        return addListener(GoogleMapEvent.PolylineRightClickEvent.class, listener);
    }
    
    public Registration addPolylinePathChangedListener(ComponentEventListener<GoogleMapEvent.PolylinePathChangedEvent> listener) {
        return addListener(GoogleMapEvent.PolylinePathChangedEvent.class, listener);
    }
    
    public Registration addVertexRightClickListener(ComponentEventListener<GoogleMapEvent.VertexRightClickEvent> listener) {
        return addListener(GoogleMapEvent.VertexRightClickEvent.class, listener);
    }
}
